### PR TITLE
Adding support for commas in active views list

### PR DESCRIPTION
**Pull request by larochj**
Added a set/get/clear/remove for the active views and displays.
Made changes to the YAML save to avoid repeating the parsing of the string list.

### CODE REVIEW

File: src/OpenColorIO/Config.cpp [hunk 1]
```diff
@@ -4005,6 +4005,69 @@ const char * Config::getActiveDisplays() const
...
+                            "non-empty name.");
+    }
+        
+    auto it = std::find(getImpl()->m_activeDisplays.begin(),
+                                   getImpl()->m_activeDisplays.end(), display);
+
+    if(it!=getImpl()->m_activeDisplays.end())
+    {
+            getImpl()->m_activeDisplays.erase(it);
+    }
```
Comment (doug-walker) [inline src/OpenColorIO/Config.cpp:4022]: I'm wondering if we should throw here? Some of the other remove*() functions throw if the name is not found.

File: src/OpenColorIO/Config.cpp [hunk 2]
```diff
@@ -4005,6 +4005,69 @@ const char * Config::getActiveDisplays() const
...
+
+void Config::addActiveDisplay(const char * view)
+{
+    if( !view || !view[0] )
+    {
+        throw Exception("Active view could not be added to config, view name has to be a "
+                        "non-empty name.");
+    }
+
+    getImpl()->m_activeDisplays.push_back(view);
```
Comment (doug-walker) [inline src/OpenColorIO/Config.cpp:4063]: We should perhaps check if it's already in the list before adding a duplicate?

File: src/OpenColorIO/Config.cpp [hunk 3]
```diff
@@ -4005,6 +4005,69 @@ const char * Config::getActiveDisplays() const
...
+
+    return getImpl()->m_activeDisplays[index].c_str();
+}
+
+int Config::getNumActiveDisplays() const
+{
+    return static_cast<int>(getImpl()->m_activeDisplays.size());
+}
+
+void Config::addActiveDisplay(const char * view)
```
Comment (doug-walker) [inline src/OpenColorIO/Config.cpp:4055]: Minor, but it would be nice if the ordering of the functions was consistent between displays and views. Perhaps move this up above removeActiveDisplay?

File: src/OpenColorIO/Config.cpp [hunk 4]
```diff
@@ -4005,6 +4005,69 @@ const char * Config::getActiveDisplays() const
...
+    AutoMutex lock(getImpl()->m_cacheidMutex);
+    getImpl()->resetCacheIDs();
+}
+
+const char * Config::getActiveDisplay( int index ) const
+{
+    if( index<0 || index >= static_cast<int>(getImpl()->m_activeDisplays.size()))
+    {
+        return nullptr;
+    }
```
Comment (doug-walker) [inline src/OpenColorIO/Config.cpp:4045]: I was wondering if we should throw, but this is fine. It's consistent with getDisplay(index).

**Review by doug-walker — CHANGES_REQUESTED** [review]
Thanks for adding this Jonathan, I think it will be very helpful for OCIO!

Please add unit tests in tests/cpu/ParseUtils_tests.cpp. This should include the various edge cases, including unmatched quotes.

File: src/OpenColorIO/ParseUtils.cpp [hunk 5]
```diff
@@ -690,33 +690,104 @@ bool StrEqualsCaseIgnore(const std::string & a, const std::string & b)
...
 
 std::string JoinStringEnvStyle(const StringUtils::StringVec & outputvec)
 {
-    return StringUtils::Join(outputvec, ',');
+    std::string result;
+    for( const auto & val : outputvec )
+    {
+        if( val.find_first_of(',') != std::string::npos )
+        {
+            result += "\"" + val + "\", ";
```
[the remark this answers is not in the document]
Comment (larochj) [reply src/OpenColorIO/ParseUtils.cpp]: Fixed, now the condition check the boundaries if there is quotes

File: src/OpenColorIO/ParseUtils.cpp [hunk 6]
```diff
@@ -690,33 +690,104 @@ bool StrEqualsCaseIgnore(const std::string & a, const std::string & b)
...
+        } 
+        else if( s[nameEndPos] == '"' )
+        {
+            // We found a quote, we need to find the next one
+            nameEndPos = s.find_first_of("\"", nameEndPos + 1);
+            if(nameEndPos == std::string::npos)
+            {
+                // We reached the end of the list instead
+                nameEndPos = s.size() - 1;
+                isEndFound = true;
```
[the remark this answers is not in the document]
Comment (larochj) [reply src/OpenColorIO/ParseUtils.cpp:719]: Right now the quote get included in the last name.
But yes I can make it throw if we don't want any name using quotes to avoid any parsing error afterward

File: src/OpenColorIO/OCIOYaml.cpp [hunk 7]
```diff
@@ -4992,13 +4992,24 @@ inline void save(YAML::Emitter & out, const Config & config)
...
     StringUtils::StringVec active_views;
-    if(config.getActiveViews() != NULL && strlen(config.getActiveViews()) > 0)
-        active_views = SplitStringEnvStyle(config.getActiveViews());
+    int nViews = config.getNumActiveViews();
+    active_views.reserve( nViews );
+    for (int i = 0; i < nViews; i++)
+    {
+        active_views.push_back(config.getActiveView(i));
+    }
     out << YAML::Value << YAML::Flow << active_views;
```
[the remark this answers is not in the document]
Comment (larochj) [reply src/OpenColorIO/OCIOYaml.cpp:5013]: Not sure if it is caused by the YAML::Flow, but yes the operator<< into a YAM::Emiter handle the quoting

File: src/OpenColorIO/ParseUtils.cpp [hunk 8]
```diff
@@ -690,33 +690,104 @@ bool StrEqualsCaseIgnore(const std::string & a, const std::string & b)
...
+    {
+        if( val.find_first_of(',') != std::string::npos )
+        {
+            result += "\"" + val + "\", ";
+        }
+        else
+        {
+            result += val + ", ";
+        }
+    }
```
[the remark this answers is not in the document]
Comment (larochj) [reply src/OpenColorIO/ParseUtils.cpp:822]: Fixed, changed it to handle the first element like  StringUtils::Join

File: src/OpenColorIO/Config.cpp (hunk 1 above)
Comment (larochj) [reply src/OpenColorIO/Config.cpp:4056]: Fixed, it throw an exception now

File: src/OpenColorIO/Config.cpp (hunk 2 above)
Comment (larochj) [reply src/OpenColorIO/Config.cpp]: Fixed, an exception is thrown on a duplicate now

File: src/OpenColorIO/Config.cpp (hunk 3 above)
Comment (larochj) [reply src/OpenColorIO/Config.cpp]: Fixed

File: src/OpenColorIO/Config.cpp (hunk 4 above)
Comment (larochj) [reply src/OpenColorIO/Config.cpp:4080]: I have left it unchanged since it fit with similar calls

File: src/OpenColorIO/ParseUtils.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (doug-walker) [reply src/OpenColorIO/ParseUtils.cpp:719]: Yes, let's throw.

File: tests/cpu/ParseUtils_tests.cpp [hunk 9]
```diff
@@ -388,37 +388,124 @@ OCIO_ADD_TEST(ParseUtils, string_vec_to_int_vec)
...
+    outputvec = { "", "", "" };
+    OCIO_CHECK_EQUAL( ", , ", OCIO::JoinStringEnvStyle(outputvec) );
+    outputvec.clear();
+
+    outputvec = { "This  : is", "a: test" };
+    OCIO_CHECK_EQUAL( "\"This  : is\", \"a: test\"", OCIO::JoinStringEnvStyle(outputvec) );
+    outputvec.clear();
+
+    outputvec = { "This", "is   \":   a:   test" };
+    OCIO_CHECK_EQUAL( "This, \"is   \":   a:   test\"", OCIO::JoinStringEnvStyle(outputvec) );
```
Comment (doug-walker) [inline tests/cpu/ParseUtils_tests.cpp:508]: Please add more tests that use commas and quotes, such as:
outputvec = { "This", "is:   ", "\"a very good,\"", " fine, helpful, and useful ", "test" };

File: src/OpenColorIO/Config.cpp [hunk 10]
```diff
@@ -4005,6 +4005,88 @@ const char * Config::getActiveDisplays() const
     return getImpl()->m_activeDisplaysStr.c_str();
 }
 
+void Config::addActiveDisplay(const char * view)
```
Comment (doug-walker) [inline src/OpenColorIO/Config.cpp:4008]: The argument to the function should be "display", not "view".

File: src/OpenColorIO/Config.cpp [hunk 11]
```diff
@@ -4005,6 +4005,88 @@ const char * Config::getActiveDisplays() const
...
+    auto it = std::find(getImpl()->m_activeDisplays.begin(),
+                        getImpl()->m_activeDisplays.end(), view);
+
+    if( it != getImpl()->m_activeDisplays.end() )
+    {
+        std::ostringstream os;
+        os << "Active display could not be added to config. An active display named '"
+           << view << "' already exists.";
+        throw Exception(os.str().c_str());
+    }
```
Comment (doug-walker) [inline src/OpenColorIO/Config.cpp:4025]: The addActiveView should have the same behavior. Please compare all of your new display functions with the view functions and make sure they behave the same.

**Review by doug-walker — CHANGES_REQUESTED** [review]

File: src/OpenColorIO/LookParse.cpp [hunk 12]
```diff
@@ -84,6 +84,12 @@ const LookParseResult::Options & LookParseResult::parse(const std::string & look
             tokens.push_back(t);
         }
 
+        if( vec.size() == 0 )
+        {
+            LookParseResult::Token t;
+            tokens.push_back(t);
+        }
```
[the remark this answers is not in the document]
Comment (larochj) [reply src/OpenColorIO/LookParse.cpp:91]: This test failure sums up why it might be needed:
```
/Users/larochj/Documents/Git/adsk-ocio-fork/tests/cpu/FileRules_tests.cpp:62:
FAILED: ss.str() == std::string(CONFIG)
	values were 'ocio_profile_version: 1

search_path: ""
strictparsing: false
luma: [0.2126, 0.7152, 0.0722]

roles:
  default: raw

displays:
  sRGB:
    - !<View> {name: Raw, colorspace: raw}

active_displays: [""]
active_views: [""]

colorspaces:
  - !<ColorSpace>
    name: raw
    family: ""
    equalitygroup: ""
    bitdepth: unknown
    isdata: false
    allocation: uniform
' and 'ocio_profile_version: 1

search_path: ""
strictparsing: false
luma: [0.2126, 0.7152, 0.0722]

roles:
  default: raw

displays:
  sRGB:
    - !<View> {name: Raw, colorspace: raw}

active_displays: []
active_views: []

colorspaces:
  - !<ColorSpace>
    name: raw
    family: ""
    equalitygroup: ""
    bitdepth: unknown
    isdata: false
    allocation: uniform
'
```

[the remark this answers is not in the document]
Comment (larochj) [reply src/OpenColorIO/LookParse.cpp:91]: We can assume an empty string to be an empty list 
Or 
an empty string to be a list containing an empty element.

For the looks it make sense to assume a list with one empty element.
For the active views and active displays, we might want to consider this an empty list.

I found an alternative to avoid changing the look parser.
Instead I let the split assume an empty string to be a list containing 1 empty element
and the getNumActiveView and display would verify the single element:
```
int Config::getNumActiveViews() const
{
    const int numActiveViews = static_cast<int>(getImpl()->m_activeViews.size());
    if( numActiveViews == 1 && 
        !getImpl()->m_activeViews[0].empty() )
    {
        return 0;
    }
    
    return static_cast<int>(getImpl()->m_activeViews.size());
}
```

We can also let users verify themselves if there is a single empty element.

File: src/OpenColorIO/Config.cpp (hunk 11 above)
Comment (larochj) [reply src/OpenColorIO/Config.cpp:4025]: Fixed

File: src/OpenColorIO/Config.cpp (hunk 10 above)
Comment (larochj) [reply src/OpenColorIO/Config.cpp]: Fixed

File: tests/cpu/ParseUtils_tests.cpp (hunk 9 above)
Comment (larochj) [reply tests/cpu/ParseUtils_tests.cpp:494]: Added this test case and another one

File: src/OpenColorIO/ParseUtils.cpp [hunk 13]
```diff
@@ -690,33 +690,138 @@ bool StrEqualsCaseIgnore(const std::string & a, const std::string & b)
...
+    if( s.size() == 0 ) 
     {
-        outputvec = StringUtils::Split(s, ',');
+        return {};
     }
-    else if (StringUtils::Find(s, ":") != std::string::npos)
-    {
-        outputvec = StringUtils::Split(s, ':');
-    }
-    else
```
[the remark this answers is not in the document]
Comment (larochj) [reply src/OpenColorIO/ParseUtils.cpp:705]: Fixed, added the separator char as a parameter

File: tests/cpu/ParseUtils_tests.cpp [hunk 14]
```diff
@@ -388,37 +388,124 @@ OCIO_ADD_TEST(ParseUtils, string_vec_to_int_vec)
...
+
     outputvec = OCIO::SplitStringEnvStyle("This:is   ,   a:test  ");
-    OCIO_CHECK_EQUAL(2, outputvec.size());
-    OCIO_CHECK_EQUAL("This:is", outputvec[0]);
-    OCIO_CHECK_EQUAL("a:test", outputvec[1]);
+    OCIO_CHECK_EQUAL(4, outputvec.size());
+    OCIO_CHECK_EQUAL("This", outputvec[0]);
+    OCIO_CHECK_EQUAL("is", outputvec[1]);
+    OCIO_CHECK_EQUAL("a", outputvec[2]);
+    OCIO_CHECK_EQUAL("test", outputvec[3]);
```
[the remark this answers is not in the document]
Comment (larochj) [reply tests/cpu/ParseUtils_tests.cpp]: Fix, it is either colon or comma

File: tests/cpu/ParseUtils_tests.cpp (hunk 9 above)
Comment (doug-walker) [reply tests/cpu/ParseUtils_tests.cpp:494]: Actually, please make the first word "\"This\"", to test that quotes are omitted when they aren't necessary.

File: tests/cpu/ParseUtils_tests.cpp [hunk 15]
```diff
@@ -388,37 +388,117 @@ OCIO_ADD_TEST(ParseUtils, string_vec_to_int_vec)
...
+    OCIO_CHECK_EQUAL("This", outputvec[0]);
+    OCIO_CHECK_EQUAL("is   \":   a:   test  \"" , outputvec[1]);
+
+    outputvec = OCIO::SplitStringEnvStyle("   \"This  : is   \",   a,   test  ");
+    OCIO_CHECK_EQUAL(3, outputvec.size());
+    OCIO_CHECK_EQUAL("This  : is   ", outputvec[0]);
+    OCIO_CHECK_EQUAL("a", outputvec[1]);
+    OCIO_CHECK_EQUAL("test", outputvec[2]);
+
+    outputvec = OCIO::SplitStringEnvStyle("   \"This  , is   \":   a:   test  ");
```
Comment (doug-walker) [inline tests/cpu/ParseUtils_tests.cpp:452]: Please add:
// If the string contains a comma, it is chosen as the separator character rather than the colon (even if it is within quotes and therefore not used as such).

File: src/OpenColorIO/ParseUtils.cpp [hunk 16]
```diff
@@ -690,33 +690,156 @@ bool StrEqualsCaseIgnore(const std::string & a, const std::string & b)
     return 0 == Platform::Strcasecmp(a.c_str(), b.c_str());
 }
 
+// Find the end of a name from a list contained in a string.
+// The element of the list are separeted by ":" or ",".
```
Comment (doug-walker) [inline src/OpenColorIO/ParseUtils.cpp:694]: Update comment.

File: src/OpenColorIO/LookParse.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (doug-walker) [reply src/OpenColorIO/LookParse.cpp:91]: Please update the getNumActive* functions to return 0 for an empty string.

However, it looks like your current SplitStringEnvStyle is already returning a StringVec with one empty string, which seems like the same thing as the previous SplitStringEnvStyle. So I guess you could have already removed this change to LookParse?

**Review by doug-walker — CHANGES_REQUESTED** [review]

File: src/OpenColorIO/LookParse.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (larochj) [reply src/OpenColorIO/LookParse.cpp]: Just fixed it, it was partly fixed at the time of your review I had found issues midway.

File: tests/cpu/ParseUtils_tests.cpp (hunk 9 above)
Comment (larochj) [reply tests/cpu/ParseUtils_tests.cpp:494]: Join would not remove the quotes in this case, but it would not add another around a string that is already surrounded by quotes. I would expect split to remove the quotes (not join).

I changed two test case to touch what I just described.

File: tests/cpu/ParseUtils_tests.cpp (hunk 15 above)
Comment (larochj) [reply tests/cpu/ParseUtils_tests.cpp:455]: Fixed

File: src/OpenColorIO/ParseUtils.cpp (hunk 16 above)
Comment (larochj) [reply src/OpenColorIO/ParseUtils.cpp]: Fixed

**Review by doug-walker — APPROVED** [review]
If you're able to fix the Windows CI failure, that would be great. Otherwise, go ahead and merge and I will fix it before submitting to OCIO.

Don't worry about the other CI failures. I need to pull over the latest Actions workflow from OCIO main to get that working again on our branch.
